src/object/cli_ec.c

```diff
@@ -2639,7 +2636,75 @@ obj_recx_ec2_daos(struct daos_oclass_attr *oca, int shard,
 		}
 	}
 
-	*output_nr = total;
-	*output_recxs = tgt_recxs;
+	D_FREE(*recxs_p);
+	*recxs_p = tgt_recxs;
+	*nr = total;
+	return 0;
+}
+
+/* Convert DAOS offset to specific data target daos offset */
+int
+obj_recx_ec_daos2shard(struct daos_oclass_attr *oca, int shard,
+		       daos_recx_t **recxs_p, unsigned int *iod_nr)
+{
+	daos_recx_t	*recx = *recxs_p;
+	int		nr = *iod_nr;
+	int		cell_nr = obj_ec_cell_rec_nr(oca);
+	int		stripe_nr = obj_ec_stripe_rec_nr(oca);
+	daos_recx_t	*tgt_recxs;
+	int		shard_idx = shard % obj_ec_tgt_nr(oca);
+	int		total;
+	int		idx;
+	int		i;
+
+	D_ASSERT(shard_idx < obj_ec_data_tgt_nr(oca));
+	for (i = 0, total = 0; i < nr; i++) {
+		uint64_t offset = recx[i].rx_idx;
+		uint64_t end = offset + recx[i].rx_nr;
+
+		while (offset < end) {
+			int tgt_idx = obj_ec_tgt_of_recx_idx(offset,
+						stripe_nr, cell_nr);
+			if (tgt_idx <= shard_idx)
+				total++;
+			offset = roundup(offset + 1, stripe_nr);
+		}
+	}
+
+	if (total == 0)
+		return 0;
+
+	D_ALLOC_ARRAY(tgt_recxs, total);
+	if (tgt_recxs == NULL)
+		return -DER_NOMEM;
+
+	for (i = 0, idx = 0; i < nr; i++) {
+		uint64_t offset = recx[i].rx_idx;
+		uint64_t end = offset + recx[i].rx_nr;
+
+		while (offset < end) {
+			int tgt_idx = obj_ec_tgt_of_recx_idx(offset,
+						stripe_nr, cell_nr);
+			if (tgt_idx <= shard_idx) {
+				uint64_t rx_idx;
+				uint64_t rx_nr;
+
+				rx_idx = max(offset,
+					     rounddown(offset, stripe_nr) +
+					     shard_idx * cell_nr);
+				rx_nr = min(roundup(rx_idx + 1, cell_nr) -
+					    rx_idx, end - offset);
+				tgt_recxs[idx].rx_idx = rx_idx;
+				tgt_recxs[idx].rx_nr = rx_nr;
+				idx++;
```

Comment:
seems still with potential problem, for example got a replicated small recx [0, 4] that should located in data tgt 0 and in parity tgts. Now rebuild data tgt 1, and iterate got recx[0, 4] from parity tgt, but this recx need not to be rebuilt on data tgt 1.
But by above logic, seems will generate an incorrect recx for data tgt 1?

Reply:
I thought the partial update in the parity node is the daos offset, instead of vos offset.  so      

int tgt_idx = obj_ec_tgt_of_recx_idx(offset, stripe_nr, cell_nr);

will get the correct target index from the offset.   so the partial update will be rebuilt on the correct target.